design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+- Moving Filters: In the example query, we currently evaluate the filter on column Y after computing the aggregation result. A worthwhile optimization for this query (which is technically valid though not a good query) would be to filter before computing the aggregation. However this would be a compatibility breaking change because at the time of an upgrade the repartition topic would contain the unfiltered rows. 
+
+#### Execution Plans
+To address this, I propose we build our own representation of the topology (called an Execution Plan) that provides details about what the streams operators are actually doing. Then, in our compatibility verification tests, we can verify that Execution Plans are kept consistent across versions.
+
+The Execution Plan is a DEG of Execution Steps. Each Execution Step represents a call to Kafka Streams, along with context about what the resulting Streams operator is doing. We can build the Execution Plan at the SchemaKStream layer.
+
+For example, the call to SchemaKStream.select would look something like:
+
+```
+public SchemaKStream<K> select(
+    final List<SelectExpression> selectExpressions,
+    final QueryContext.Stacker contextStacker,
+    final ProcessingLogContext processingLogContext) {
+  final ExecutionStep step = new StreamSelect(this.executionStep, selectExpressions);
+  return new SchemaKStream<>(
+      step.build(),
+      ksqlConfig,
+      functionRegistry,
+      contextStacker.getQueryContext(),
+      step
+  );
+}
+```
+
+Where StreamSelect looks like:
+
+```
+public class StreamSelect implements ExecutionStep {
+  ...
+  private final ExecutionStep source;
+  private final List<SelectExpression> selectExpressions;
+
+  @JsonCreator
+  public StreamSelect(
+      ...
+      @JsonProperty("source") final ExecutionStep source,
+      @JsonProperty("selectExpressions") final List<SelectExpression> selectExpressions) {
+    this.properties = Objects.requireNonNull(properties, "properties");
+    this.source = Objects.requireNonNull(source, "source");
+    this.selectExpressions = ImmutableList.copyOf(selectExpressions);
+  }
+
+  public List<SelectExpression> getSelectExpressions() {
+    return selectExpressions;
+  }
+
+  @Override
+  public List<ExecutionStep> getSources() {
+    return Collections.singletonList(source);
+  }
+
+  @Override
+  public KStream build(final KStream input) {
+      final Selection selection = new Selection(selectExpressions);
+      return kstream.mapValues(selection.getSelectValueMapper()),
+  }
+}
+```
+
+I've written up a basic implementation of Execution Plans which includes all the Execution Steps we'd need, and can be found here (https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java) (steps are under ksql-engine/src/main/java/io/confluent/ksql/structured/execution):
```

Comment:
Sorry, where in `https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java` am I supposed to be looking...?

Reply:
Oops, that's the wrong file.